linera-rpc/src/grpc/server.rs

```diff
@@ -399,75 +400,133 @@ where
         cross_chain_sender_failure_rate: f32,
         cross_chain_max_concurrent_tasks: usize,
         this_shard: ShardId,
-        receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
+        mut receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
     ) {
         let pool = GrpcConnectionPool::default();
-        let max_concurrent_tasks = Some(cross_chain_max_concurrent_tasks);
-
-        receiver
-            .for_each_concurrent(max_concurrent_tasks, |(cross_chain_request, shard_id)| {
-                let shard = network.shard(shard_id);
-                let remote_address = shard.http_address();
-
-                let pool = pool.clone();
-                let nickname = nickname.clone();
-
+        let mut join_set = JoinSet::new();
+        // An entry means that a task is running for the given sender, recipient and request kind.
+        // If it is Some, that means the task is still working on another request, and will pick
```

Comment:
As a rule of thumb, if I have to use comments to attach meaning to `Option` I usually take it as an indication that I should create a custom enum for the task.

Reply:
Do you have an idea for names?
Something like:

```rust
enum CrossChainTaskQueue {
    Empty,
    Full(CrossChainRequest),
}
```

Reply:
(Not really a "queue", if it's only one task…)

Reply:
Maybe it's just a ‘pending task’?

Alternatively, since empty is never a valid state for the task, we could remove this option and use the empty state for what's currently `None`?